src/docker/bitcoin.rs

```diff
@@ -0,0 +1,244 @@
+use crate::docker::{
+    self, free_local_port::free_local_port, DockerImage, LogMessage, DOCKER_NETWORK,
+};
+use anyhow::Context;
+use futures::compat::Future01CompatExt;
+use reqwest::r#async::Client;
+use rust_bitcoin::{
+    self,
+    hashes::sha256d,
+    util::bip32::{ChildNumber, ExtendedPrivKey},
+    Address, Amount, Network,
+};
+use secp256k1::{
+    rand::{thread_rng, Rng},
+    Secp256k1,
+};
+use shiplift::ContainerOptions;
+
+const IMAGE: &str = "coblox/bitcoin-core:0.17.0";
+
+const USERNAME: &str = "bitcoin";
+const PASSWORD: &str = "t68ej4UX2pB0cLlGwSwHFBLKxXYgomkXyFyxuBmm2U8=";
+
+#[derive(derive_more::Display, Copy, Clone)]
+#[display(fmt = "127.0.0.1:{}", port)]
+pub struct BitcoindP2PUri {
+    port: u16,
+}
+
+#[derive(derive_more::Display, Copy, Clone)]
+#[display(fmt = "http://localhost:{}", port)]
+pub struct BitcoindHttpEndpoint {
+    port: u16,
+}
+
+pub struct BitcoindInstance {
+    pub p2p_uri: BitcoindP2PUri,
+    pub http_endpoint: BitcoindHttpEndpoint,
+    pub account_0: Account,
+    pub account_1: Account,
+}
+
+pub async fn new_bitcoind_instance() -> anyhow::Result<BitcoindInstance> {
+    let mut options_builder = ContainerOptions::builder(IMAGE);
+    options_builder.name("bitcoin");
+    options_builder.network_mode(DOCKER_NETWORK);
+    options_builder.cmd(vec![
+        "-regtest",
+        "-server",
+        "-rest",
+        "-printtoconsole",
+        "-bind=0.0.0.0:18444",
+        "-rpcbind=0.0.0.0:18443",
+        "-rpcauth=bitcoin:1c0e8f3de84926c04115e7da7e501346$a48f42ad32741dd1755649c8b98663b3ccbebeb75f196389f9a5c8a96b72edb3",
+        "-rpcallowip=0.0.0.0/0",
+        "-debug=1",
+        "-acceptnonstdtxn=0",
+        "-txindex",
+    ]);
+
+    let p2p_port = free_local_port().await?;
+    options_builder.expose(18444, "tcp", p2p_port as u32);
+
+    let p2p_uri = BitcoindP2PUri { port: p2p_port };
+
+    let http_port = free_local_port().await?;
+    options_builder.expose(18443, "tcp", http_port as u32);
+
+    let http_endpoint = BitcoindHttpEndpoint { port: http_port };
+
+    let options = options_builder.build();
+
+    docker::start(
+        DockerImage(IMAGE),
+        options,
+        LogMessage("Flushed wallet.dat"),
+    )
+    .await?;
+
```

Comment:
This has a race condition.  When `free_local_port()` returns the socket is closed.  If another process opens a socket it may be on the port returned by `free_local_port()` and then docker will not be able to successfully start.  I cannot think of a way to solve this without burdening the user to configure the port number and the race condition is probably not a big deal here, might be worth noting it in a comment.

Reply:
Well spotted!

Apparently, there is logic in the kernel(s) that still makes this work. To properly fix it, we would have to delegate the port assignment to docker by passing 0 and then read using docker inspect, which port we got (that is actually what we do in `testcontainers` :) )